### ISSUE TITLE
Improving UQ_KPP

### ISSUE THREAD

**Issue by glwagner** [glwagner/Dao.jl#1]
A few suggestions:

1. Separate the analysis of the Oceananignas NetCDF output and the UQ. In particular, let's create an array of 1D profiles of the LES data and save them separately. I propose we use 

https://github.com/JuliaIO/JLD2.jl

2. The first processing step should be to convert the 1D JLD2 data from the LES to OceanTurb fields.

3. We could initialize the OceanTurb profile with the LES data?

4. I couldn't figure out why the parameters have to be stored in an array, as here:

```julia
params = [0.1 0.4 6.33 2.0 6.4 0.599 0.135 1.36 -1.85 0.5 2.5 4.32 0.3 1e-11]
parameters = KPP.Parameters(
    Cε = params[1], Cκ = params[2], CN = params[3], Cstab = params[4], 
    Cunst = params[5], Cb_U = params[6], Cτ_U = params[7], Cb_T  = params[8],
    Cτ_T = params[9], Cd_U = params[10], Cd_T = params[11], CRi = params[12], CKE = params[13]
)
```

Also, since this just uses defaults, its no different from writing

```julia
parameters = KPP.Parameters()
```
which is easier, presumably.

5. To 'splat' parameters we can do

```julia
params_dict = Dict(:Cstab => 2.1)
parameters = KPP.Parameters(; params_dict...)
```

or something like that.

But I think what we really need is some pseudocode to indicate what we are trying to do? Then we can figure out the julia algorithm to do it.

6. Let's use this function (for example):

```julia
function free_convection_model(N, L, parameters, heat_flux, γ)
    model = KPP.Model(N=N, L=L, parameters=parameters)
    T₀(z) = 20 + γ*z
    model.solution.T = T₀
    temperature_flux = heat_flux / (model.constants.ρ₀ * model.constants.cP)

    model.bcs.T.top = FluxBoundaryCondition(temperature_flux)
    model.bcs.T.bottom = GradientBoundaryCondition(γ)

    return model
end
```

6. Finally, let's modify the default `Constants` in OceanTurb to be identical to `Oceananigans`.

**Comment by ali-ramadhan** [comment]
> 6. Finally, let's modify the default `Constants` in OceanTurb to be identical to `Oceananigans`.

Might be good to even decide on a shared abstraction/structure. The one used by Oceananigans.jl was just thrown together a while back, might be good to revisit.

**Comment by glwagner** [comment]
Let's discuss this block of code:

```julia
    ...
    T_p = []
    tmp1 = zeros(128)
    tmp = data(model.solution.T)
    @. tmp1 = tmp
    push!(T_p,tmp1[end:-1:1])

    reset!(model.clock)
    for i in 1:ni
        iterate!(model, dt, nt)
        tmp = data(model.solution.T)
        @. tmp1 = tmp
        push!(T_p,tmp1[end:-1:1])
    end
    loss = 0
    for i in 1:(ni+1)
        loss += mean((T_p[i] .- T_s[i] ).^2 ) / mean(T_s[i] .^2)
    end
    loss *= 1 / (ni+1)
    loss==loss ? loss = loss : loss = Inf
    return loss
end
```

**Comment by glwagner** [comment]
What I think is happening is

1. An empty array called `T_p` is created --- this is approximate model solution? It might be better to call these variables `T_model` and `T_data` --- or something like that.

2. `tmp` doesn't ever appear to be used, except to be copied to `tmp1`. Why do we need `tmp`?

3. `tmp1` is reversed to match the `Oceananigans` convention. But perhaps it'd be better to pre-process the `Oceananigans` data?

4. It looks like the loss function could be written

```julia
norm(T_model - T_data)^2 / norm(T_data)^2
```

right?

5. Let's clean up the last line. I think it'd be clearer to write

```julia
if isnan(loss)
    loss = Inf
end
```

**Comment by glwagner** [comment]
> Might be good to even decide on a shared abstraction/structure. The one used by Oceananigans.jl was just thrown together a while back, might be good to revisit.

@ali-ramadhan we could do that! 

We could also just write a function in `OceanTurb` that can interpret the `Oceananigans` data structure, so we have something like

```julia
constants = Constants(Oceananignas.EquationOfState())
```

or whatever it needs to be...

**Comment by ali-ramadhan** [comment]
That works too! We can always work on a nicer abstraction down the line after we have a better idea of what we want.

**Comment by glwagner** [comment]
I think a simple struct works well. I guess the important thing is deciding on names for the variables, and which variables to include.
